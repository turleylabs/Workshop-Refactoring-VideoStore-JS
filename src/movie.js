class Movie {

  constructor(title, priceCode) {
    this.title = title;
    this.priceCode = priceCode;
  }

  getPriceCode() {
    return this.priceCode;
  }
  
  getTitle() {
    return this.title;
  };
}

const CHILDRENS_MOVIE = 2;
const REGULAR_MOVIE = 0;
const NEW_RELEASE_MOVIE = 1;

export {Movie, CHILDRENS_MOVIE, REGULAR_MOVIE, NEW_RELEASE_MOVIE};