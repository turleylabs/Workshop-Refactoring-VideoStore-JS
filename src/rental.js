class Rental {
  constructor(movie, daysRented) {
      this.movie = movie;
      this.daysRented = daysRented;
  }

  getDaysRented() {
      return this.daysRented;
  }

  getMovie() {
      return this.movie;
  }

}

export {Rental};