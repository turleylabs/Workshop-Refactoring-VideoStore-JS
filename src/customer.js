import {REGULAR_MOVIE, CHILDRENS_MOVIE, NEW_RELEASE_MOVIE} from './movie';

class Customer {
  constructor(name) {
      this.name = name;
      this.rentals = [];
  }

  addRental(rental) {
    this.rentals.push(rental);
  }

  createStatement() {
      let totalAmount = 0.0;
      let frequentRenterPoints = 0;
      let result = "Rental Record for " + this.name + "\n";
      for (const each of this.rentals) {
          let thisAmount = 0;

          //determine amounts for each line
          switch (each.getMovie().getPriceCode()) {
              case REGULAR_MOVIE:
                  thisAmount += 2;
                  if (each.getDaysRented() > 2)
                      thisAmount += (each.getDaysRented() - 2) * 1.5;
                  break;
              case NEW_RELEASE_MOVIE:
                  thisAmount += each.getDaysRented() * 3;
                  break;
              case CHILDRENS_MOVIE:
                  thisAmount += 1.5;
                  if (each.getDaysRented() > 3)
                      thisAmount += (each.getDaysRented() - 3) * 1.5;
                  break;
          }

          frequentRenterPoints++;
          if ((each.getMovie().getPriceCode() == NEW_RELEASE_MOVIE) &&
                  each.getDaysRented() > 1) frequentRenterPoints++;

          result += "\t" + each.getMovie().getTitle() + "\t" +
                  thisAmount + "\n";
          totalAmount += thisAmount;

      }
      result += "Amount owed is " + totalAmount + "\n";
      result += "You earned " + frequentRenterPoints +
              " frequent renter points";
      return result;
  }

}

export {Customer};