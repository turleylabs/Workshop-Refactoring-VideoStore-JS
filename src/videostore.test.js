import product from 'cartesian-product';
import { Customer } from './customer';
import { CHILDRENS_MOVIE, Movie, NEW_RELEASE_MOVIE, REGULAR_MOVIE } from './movie';
import { Rental } from './rental';

test('approve createStatement', () => {
    function doCreateStatement(priceCode, daysRented) {
        const customer = new Customer("Steve");
        customer.addRental(new Rental(new Movie("Star Wars", priceCode), daysRented));
        customer.addRental(new Rental(new Movie("Raiders of the Lost Ark", REGULAR_MOVIE), 1));
        return customer.createStatement();
    }

    const priceCodes = [REGULAR_MOVIE, NEW_RELEASE_MOVIE, CHILDRENS_MOVIE, 3];
    const daysRented = [0,1,2,3,4];
    const params = [priceCodes, daysRented];

    const result = runCombinationSnapshot(params, doCreateStatement);

    expect(result).toMatchSnapshot();
});

function runCombinationSnapshot(params, fn) {
    return product(params).map(combo => [combo, fn(...combo)]);
}

//    
