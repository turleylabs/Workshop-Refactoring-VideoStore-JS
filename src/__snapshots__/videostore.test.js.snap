// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`approve createStatement 1`] = `
Array [
  Array [
    Array [
      0,
      0,
    ],
    "Rental Record for Steve
	Star Wars	2
	Raiders of the Lost Ark	2
Amount owed is 4
You earned 2 frequent renter points",
  ],
  Array [
    Array [
      0,
      1,
    ],
    "Rental Record for Steve
	Star Wars	2
	Raiders of the Lost Ark	2
Amount owed is 4
You earned 2 frequent renter points",
  ],
  Array [
    Array [
      0,
      2,
    ],
    "Rental Record for Steve
	Star Wars	2
	Raiders of the Lost Ark	2
Amount owed is 4
You earned 2 frequent renter points",
  ],
  Array [
    Array [
      0,
      3,
    ],
    "Rental Record for Steve
	Star Wars	3.5
	Raiders of the Lost Ark	2
Amount owed is 5.5
You earned 2 frequent renter points",
  ],
  Array [
    Array [
      0,
      4,
    ],
    "Rental Record for Steve
	Star Wars	5
	Raiders of the Lost Ark	2
Amount owed is 7
You earned 2 frequent renter points",
  ],
  Array [
    Array [
      1,
      0,
    ],
    "Rental Record for Steve
	Star Wars	0
	Raiders of the Lost Ark	2
Amount owed is 2
You earned 2 frequent renter points",
  ],
  Array [
    Array [
      1,
      1,
    ],
    "Rental Record for Steve
	Star Wars	3
	Raiders of the Lost Ark	2
Amount owed is 5
You earned 2 frequent renter points",
  ],
  Array [
    Array [
      1,
      2,
    ],
    "Rental Record for Steve
	Star Wars	6
	Raiders of the Lost Ark	2
Amount owed is 8
You earned 3 frequent renter points",
  ],
  Array [
    Array [
      1,
      3,
    ],
    "Rental Record for Steve
	Star Wars	9
	Raiders of the Lost Ark	2
Amount owed is 11
You earned 3 frequent renter points",
  ],
  Array [
    Array [
      1,
      4,
    ],
    "Rental Record for Steve
	Star Wars	12
	Raiders of the Lost Ark	2
Amount owed is 14
You earned 3 frequent renter points",
  ],
  Array [
    Array [
      2,
      0,
    ],
    "Rental Record for Steve
	Star Wars	1.5
	Raiders of the Lost Ark	2
Amount owed is 3.5
You earned 2 frequent renter points",
  ],
  Array [
    Array [
      2,
      1,
    ],
    "Rental Record for Steve
	Star Wars	1.5
	Raiders of the Lost Ark	2
Amount owed is 3.5
You earned 2 frequent renter points",
  ],
  Array [
    Array [
      2,
      2,
    ],
    "Rental Record for Steve
	Star Wars	1.5
	Raiders of the Lost Ark	2
Amount owed is 3.5
You earned 2 frequent renter points",
  ],
  Array [
    Array [
      2,
      3,
    ],
    "Rental Record for Steve
	Star Wars	1.5
	Raiders of the Lost Ark	2
Amount owed is 3.5
You earned 2 frequent renter points",
  ],
  Array [
    Array [
      2,
      4,
    ],
    "Rental Record for Steve
	Star Wars	3
	Raiders of the Lost Ark	2
Amount owed is 5
You earned 2 frequent renter points",
  ],
  Array [
    Array [
      3,
      0,
    ],
    "Rental Record for Steve
	Star Wars	0
	Raiders of the Lost Ark	2
Amount owed is 2
You earned 2 frequent renter points",
  ],
  Array [
    Array [
      3,
      1,
    ],
    "Rental Record for Steve
	Star Wars	0
	Raiders of the Lost Ark	2
Amount owed is 2
You earned 2 frequent renter points",
  ],
  Array [
    Array [
      3,
      2,
    ],
    "Rental Record for Steve
	Star Wars	0
	Raiders of the Lost Ark	2
Amount owed is 2
You earned 2 frequent renter points",
  ],
  Array [
    Array [
      3,
      3,
    ],
    "Rental Record for Steve
	Star Wars	0
	Raiders of the Lost Ark	2
Amount owed is 2
You earned 2 frequent renter points",
  ],
  Array [
    Array [
      3,
      4,
    ],
    "Rental Record for Steve
	Star Wars	0
	Raiders of the Lost Ark	2
Amount owed is 2
You earned 2 frequent renter points",
  ],
]
`;
